packages/@react-aria/landmark/src/useLandmark.ts

```diff
@@ -0,0 +1,225 @@
+import {useLayoutEffect} from '@react-aria/utils';
+
+export interface AriaLandmarkProps  extends AriaLabelingProps {
+  role: 'main' | 'region' | 'search' | 'navigation' | 'form' | 'banner' | 'contentinfo' | 'complementary'
+}
+
+interface LandmarkAria {
+  landmarkProps: HTMLAttributes<HTMLElement>
+}
+
+let landmarkRoles = ['main', 'region', 'search', 'navigation', 'form', 'banner', 'contentinfo', 'complementary'];
+const LANDMARK_ELEMENT_SELECTOR = landmarkRoles.map(value => `[role=${value}]`).join(',');
+
+function getLandmarkTreeWalker(root: HTMLElement, opts?) {
+  let selector = LANDMARK_ELEMENT_SELECTOR;
+  let walker = document.createTreeWalker(
+    root,
+    NodeFilter.SHOW_ELEMENT,
+    {
+      acceptNode(node) {
+        // Not sure we need this, can landmarks be nested?
+        // Skip nodes inside the starting node.
+        // if (opts?.from?.contains(node)) {
+        //   return NodeFilter.FILTER_REJECT;
+        // }
+
+        if ((node as HTMLElement).matches(selector)
+          && isElementVisible(node as HTMLElement)
+          && opts?.registeredLandmarks.includes(node)) {
+          return NodeFilter.FILTER_ACCEPT;
+        }
+
+        return NodeFilter.FILTER_SKIP;
+      }
+    }
+  );
+
+  if (opts?.from) {
+    walker.currentNode = opts.from;
+  }
+
+  return walker;
+}
+
+function treeDistance(child: HTMLElement, target: HTMLElement): number {
+  let distance = 0;
+  let node = child;
+  while (target && node.parentElement && node !== target) {
+    node = node.parentElement;
+    distance++;
+  }
+  return distance;
+}
+
+function hasLabel(ref: MutableRefObject<HTMLElement>) {
+  return ref.current.getAttribute('aria-label') !== null  || ref.current.getAttribute('aria-labelledby') !== null;
+}
+
+class LandmarkManager {
+  private landmarks = new Map<MutableRefObject<HTMLElement>, HTMLElement>();
+  private static instance: LandmarkManager;
+
+  private constructor() {}
+
+  public static getInstance(): LandmarkManager {
+    if (!LandmarkManager.instance) {
+      LandmarkManager.instance = new LandmarkManager();
+    }
+
+    return LandmarkManager.instance;
+  }
+
+  // Get list of landmarks with a specific role.
+  private getLandmarksByRole(role) {
+    return [...this.landmarks.keys()].filter((l) => l.current.getAttribute('role') === role);
+  }
+
+  public addLandmark(ref: MutableRefObject<HTMLElement>) {
+    // could change to a map to track roles, then we could warn if someone provides two of the same role but no labels for them
+    if (!this.landmarks.has(ref)) {
+      this.landmarks.set(ref, null);
+    }
+
+    // Warn if there are 2+ landmarks with the same role but no label
+    // https://www.w3.org/TR/wai-aria-practices/examples/landmarks/navigation.html
+    let role = ref.current.getAttribute('role');
+    let duplicateRoleLandmarks = this.getLandmarksByRole(role);
+    if (duplicateRoleLandmarks.length >= 2 && duplicateRoleLandmarks.some(r => !hasLabel(r))) {
```

Comment:
Two questions:

1. Is this how we want to do this check? (Add landmark, then console warn for missing label)
2. Should we also add a check for duplicate labels here? They should be unique.

Reply:
Generally we make these kinds of warnings in the hook, best if we could keep it there

maybe addLandmark could return the number of landmarks with the role we're trying to add, then, if the new landmark being added doesn't have some sort of label on props, then we warn

I don't really have a solid reason for this, it just feels more right to look in props instead of dom attributes. Though it'd be harder to check the first landmark added, which is why i assume you did it this way.
Maybe we should have an update function for landmarks that would allow us to add a label at a later date? We'd need to track the labels in the Map though